src/workerd/util/weak-refs.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <kj/mutex.h>
+#include <kj/refcount.h>
+
+namespace workerd {
+
+// T must itself extend from kj::AtomicRefcounted
+template <typename T>
+class AtomicWeakRef final: public kj::AtomicRefcounted {
+public:
+  inline static kj::Own<const AtomicWeakRef<T>> wrap(T* this_) {
+    return kj::atomicRefcounted<AtomicWeakRef<T>>(this_);
+  }
+
+  inline explicit AtomicWeakRef(T* thisArg) : this_(thisArg) {}
+
+  // This tries to materialize a strong reference to the owner. It will fail if the owner's
+  // refcount has already dropped to 0. As discussed in the class, the lifetime of this weak
+  // reference can exceed the lifetime of the isolate it's tracking.
```

Comment:
Since we may use AtomicWeakRef for more than just the Isolate class, let's change isolate to object or something else here.